// drivers/pg.js
"use strict";
var baseDriver = require('driver');
var pg = require('pg');
var names = {};
var uniquifier = 0;

module.exports = function (options) {
    var driver = baseDriver({
        run : function(sql, params, rowCallback, endCallback, errorCallback) {
            pg.connect(options, function(err, client) {
                var name;
                if (err) {
                    if (errorCallback !== undefined) errorCallback(err);
                } else {
                    if (names[sql] !== undefined) {
                        name = names[sql];
                    } else {
                        name = 'query' + uniquifier++;
                        names[sql] = name;
                    }
                    var q = client.query({
                        name : name,
                        text : sql,
                        values : params
                    });
                    if (rowCallback !== undefined) q.on('row', rowCallback);
                    if (endCallback !== undefined) q.on('end', endCallback);
                }
            });
        },
        runInsert : function(sql, params, returning, rowCallback, endCallback, errorCallback) {
            this.run(sql + ' RETURNING ' + returning.map(this.quoteUnqualifiedIdentifier).join(','),
                params, rowCallback, endCallback, errorCallback);
        },
        getBind : function(params) {
            if (params.bind === undefined) params.bind = 1;
            return '$' + params.bind++;
        },
        getDefaultValues : function() {
            return 'DEFAULT VALUES';
        }
    });
    return driver;
};
